src/text/line.rs

```diff
@@ -585,36 +555,96 @@ impl Widget for Line<'_> {
+        let line_width = self.width();
+        if area.is_empty() || line_width == 0 {
+            return;
+        }
         buf.set_style(area, self.style);
-        let width = self.width() as u16;
-        let mut x = area.left();
-        let line = if width > area.width {
-            self.truncated(area.width)
+
+        let area_width = usize::from(area.width);
+        let can_render_complete_line = line_width <= area_width;
+        if can_render_complete_line {
+            let indent_width = match self.alignment {
+                Some(Alignment::Center) => (area_width.saturating_sub(line_width)) / 2,
+                Some(Alignment::Right) => area_width.saturating_sub(line_width),
+                Some(Alignment::Left) | None => 0,
+            };
+            let indent_width = u16::try_from(indent_width).unwrap_or(u16::MAX);
+            let area = area.indent_x(indent_width);
+            render_spans(&self.spans, area, buf, 0);
         } else {
-            let offset = match self.alignment {
-                Some(Alignment::Center) => (area.width.saturating_sub(width)) / 2,
-                Some(Alignment::Right) => area.width.saturating_sub(width),
+            // There is not enough space to render the whole line. As the right side is truncated by
+            // the area width, only truncate the left.
+            let skip_width = match self.alignment {
+                Some(Alignment::Center) => (line_width.saturating_sub(area_width)) / 2,
+                Some(Alignment::Right) => line_width.saturating_sub(area_width),
                 Some(Alignment::Left) | None => 0,
             };
-            x = x.saturating_add(offset);
-            self.to_owned()
+            render_spans(&self.spans, area, buf, skip_width);
         };
-        for span in &line.spans {
-            let span_width = span.width() as u16;
-            let span_area = Rect {
-                x,
-                width: span_width.min(area.right().saturating_sub(x)),
-                ..area
-            };
-            span.render(span_area, buf);
-            x = x.saturating_add(span_width);
-            if x >= area.right() {
-                break;
-            }
+    }
+}
+
+/// Renders all the spans of the line that should be visible.
+fn render_spans(spans: &[Span], mut area: Rect, buf: &mut Buffer, span_skip_width: usize) {
+    for (span, span_width, offset) in visible_spans(spans, span_skip_width) {
+        area = area.indent_x(offset);
+        if area.is_empty() {
+            break;
         }
+        span.render_ref(area, buf);
+        let span_width = u16::try_from(span_width).unwrap_or(u16::MAX);
+        area = area.indent_x(span_width);
     }
 }
 
+/// Returns an iterator over the spans that are visible
+///
+/// Visibility is determined by the spans that are partially or completely visible taking into
+/// account just the amount of width to skip from the start of the `Line`.
+fn visible_spans<'a>(
```

Comment:
I don't think moving this into its own method is very useful. Its very specific to the render_span method.

Also, this was a pitfall for me in the past, this is not visibility. This is only skipping from the left. The right is still untouched and will only be trucated at the actual render. Visible would mean that everything can be u16, but that is not the case here. This should not be named `visible` as a function name is what is read first. It gives a very wrong impression about what this function is and for what it might be used in other cases.

Also, the doc comment about `partially or completely visible` is somewhat fine for code inside `render_spans` but very wrong for its own method with its own assumptions.

When this method is named maybe `skip_spans_by_width` it would be better, but it should not be named visible. Personally I would prefer it inside `render_spans` again as its very specific to the logic of `render_spans`. Otherwise state in the doc comments that this is very dependent on the assumptions of `Widget::render_ref` and move `render_spans` back into `render_ref`.

```suggestion
/// Returns an iterator over the spans that skips are visible
///
/// Visibility is determined by the spans that are partially or completely visible taking into
/// account just the amount of width to skip from the start of the `Line`.
fn skip_spans_by_width<'a>(
```